File: packages/core/src/formats.ts
import { ToolError, causeMessage } from "./errors.ts"
import { META_TOOL_NAMES } from "./meta-tools.ts"
import { EXPORT_PREFIX } from "./validate.ts"
import type {
  AnthropicTool,
  CallMeta,
  Flint,
  GeminiSchema,
  GeminiTool,
  JsonSchema,
  JsonValue,
  Manifest,
  McpTool,
  ObjectSchema,
  OpenAiChatTool,
  OpenAiTool,
  ProviderSchema,
  ToolAnnotations,
  ToolDefinition,
  ToolFormat,
  ToolFormats,
  ToolResults,
  VercelTool,
} from "./types.ts"

export { EXPORT_PREFIX }

// Gemini's Schema is a subset of OpenAPI 3.03: it has no `additionalProperties`, and its examples field is singular.
const GEMINI_UNSUPPORTED = new Set(["additionalProperties", "examples"])

// A genai Schema spells these four as strings, the way proto JSON spells an int64. `maximum` and `minimum` stay numbers.
const GEMINI_COUNTS = new Set(["minItems", "maxItems", "minLength", "maxLength"])

// Structured outputs names the keywords it takes, and these five of flintd's are not among them.
const STRICT_UNSUPPORTED = new Set(["minLength", "maxLength", "title", "default", "examples"])

// "Objects have limitations on nesting depth and size": up to 10 levels. flintd's own bound is 32.
const STRICT_MAX_DEPTH = 10

// Every meta tool but tool_run answers a JSON object of its own; tool_run answers whatever the Tool it ran answered.
const OBJECT_RESULT: ObjectSchema = { type: "object" }

const READ_ONLY_META = new Set(["tool_read", "tool_find", "tool_history"])

export function exportName(name: string): string {
  return META_TOOL_NAMES.has(name) ? name : `${EXPORT_PREFIX}${name}`
}

// `fl_` is reserved at create, so a Tool never carries it and the prefix can only be the spelling a format exports.
export function libraryName(exported: string): string {
  if (META_TOOL_NAMES.has(exported) || !exported.startsWith(EXPORT_PREFIX)) return exported
  const bare = exported.slice(EXPORT_PREFIX.length)
  return META_TOOL_NAMES.has(bare) ? exported : bare
}

export function formatTools(tools: ToolDefinition[], format: "anthropic"): AnthropicTool[]
export function formatTools(tools: ToolDefinition[], format: "openai"): OpenAiTool[]
export function formatTools(tools: ToolDefinition[], format: "openai-chat"): OpenAiChatTool[]
export function formatTools(tools: ToolDefinition[], format: "gemini"): GeminiTool[]
export function formatTools(tools: ToolDefinition[], format: "vercel"): Record<string, VercelTool>
export function formatTools(
  tools: ToolDefinition[],
  format: "mcp",
  manifests?: ReadonlyMap<string, Manifest>,
): McpTool[]
export function formatTools<F extends ToolFormat>(tools: ToolDefinition[], format: F): ToolFormats[F]
export function formatTools(
  tools: ToolDefinition[],
  format: ToolFormat,
  manifests?: ReadonlyMap<string, Manifest>,
): unknown {
  switch (format) {
    case "anthropic":
      return tools.map((tool) => ({
        name: exportName(tool.name),
        description: tool.description,
        input_schema: objectSchema(tool.parameters),
      }))
    case "openai":
      return tools.map((tool) => ({
        type: "function" as const,
        name: exportName(tool.name),
        description: tool.description,
        parameters: open(tool.parameters),
        strict: strictly(tool.parameters),
      }))
    case "openai-chat":
      return tools.map((tool) => ({
        type: "function" as const,
        function: {
          name: exportName(tool.name),
          description: tool.description,
          parameters: open(tool.parameters),
          strict: strictly(tool.parameters),
        },
      }))
    case "gemini":
      return tools.map((tool) => ({
        name: exportName(tool.name),
        description: tool.description,
        parameters: subset(tool.parameters),
        ...(tool.result === undefined ? {} : { response: subset(tool.result) }),
      }))
    case "vercel":
      return Object.fromEntries(
        tools.map((tool) => [
          exportName(tool.name),
          {
            description: tool.description,
            inputSchema: open(tool.parameters),
            ...(tool.result === undefined ? {} : { outputSchema: open(tool.result) }),
          },
        ]),
      )
    case "mcp":
      return tools.map((tool) => describe(tool, manifests?.get(tool.name)))
  }
}

// The whole MCP tool shape, so the daemon's surface and this format cannot say two different things about one Tool.
function describe(tool: ToolDefinition, manifest: Manifest | undefined): McpTool {
  const meta = META_TOOL_NAMES.has(tool.name)
  // MCP pairs an output schema with `structuredContent`, which is an object, so a result schema with any other root declares none.
  const output = meta ? (tool.name === "tool_run" ? undefined : OBJECT_RESULT) : objectRoot(tool.result)
  return {
    name: exportName(tool.name),
    description: tool.description,
    inputSchema: objectSchema(tool.parameters),
    ...(output === undefined ? {} : { outputSchema: output }),
    ...(meta ? metaAnnotations(tool.name) : { annotations: annotations(manifest ?? {}) }),
  }
}

function metaAnnotations(name: string): { annotations?: ToolAnnotations } {
  if (READ_ONLY_META.has(name)) {
    return { annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: false, openWorldHint: false } }
  }
  // tool_run carries the annotations of whatever it dispatches to, which flintd cannot know here, so it carries none.
  if (name === "tool_run") return {}
  return { annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: false, openWorldHint: false } }
}

function annotations(manifest: Manifest): ToolAnnotations {
  const files = manifest.fs !== undefined
  const commands = manifest.exec === true
  const reaches = (manifest.hosts?.length ?? 0) > 0 || (manifest.connections?.length ?? 0) > 0
  return {
    readOnlyHint: !files && !commands && !reaches,
    destructiveHint: files || commands,
    // A Body is arbitrary JavaScript, so flintd never promises that calling it twice is the same as calling it once.
    idempotentHint: false,
    openWorldHint: reaches,
  }
}

function objectRoot(schema: JsonSchema | undefined): ObjectSchema | undefined {
  return schema?.type === "object" ? objectSchema(schema) : undefined
}

export async function callFrom<F extends ToolFormat>(
  flint: Flint,
  format: F,
  block: unknown,
  meta: CallMeta = {},
): Promise<ToolResults[F]> {
  const asked = read(format, block)
  let args: JsonValue
  try {
    args = parsed(asked.args)
  } catch (cause) {
    return failure(format, asked, refusal(cause)) as ToolResults[F]
  }
  try {
    return success(format, asked, await flint.call(libraryName(asked.name), args, meta)) as ToolResults[F]
  } catch (cause) {
    return failure(format, asked, refusal(cause)) as ToolResults[F]
  }
}

interface Asked {
  id: string
  name: string
  args: unknown
}

function read(format: ToolFormat, block: unknown): Asked {
  const one = object(block, "the tool call")
  switch (format) {
    case "anthropic":
      return { id: string(one["id"], "id"), name: string(one["name"], "name"), args: one["input"] }
    case "openai":
      return { id: string(one["call_id"], "call_id"), name: string(one["name"], "name"), args: one["arguments"] }
    case "openai-chat": {
      const call = object(one["function"], "function")
      return { id: string(one["id"], "id"), name: string(call["name"], "name"), args: call["arguments"] }
    }
    case "gemini": {
      const part = "functionCall" in one ? object(one["functionCall"], "functionCall") : one
      const id = part["id"]
      return { id: id === undefined ? "" : string(id, "id"), name: string(part["name"], "name"), args: part["args"] }
    }
    case "vercel":
      return {
        id: string(one["toolCallId"], "toolCallId"),
        name: string(one["toolName"], "toolName"),
        args: one["input"],
      }
    case "mcp":
      return { id: "", name: string(one["name"], "name"), args: one["arguments"] }
  }
}

function success(format: ToolFormat, asked: Asked, result: JsonValue): ToolResults[ToolFormat] {
  switch (format) {
    case "anthropic":
      return { type: "tool_result", tool_use_id: asked.id, content: JSON.stringify(result) }
    case "openai":
      return { type: "function_call_output", call_id: asked.id, output: JSON.stringify(result) }
    case "openai-chat":
      return { role: "tool", tool_call_id: asked.id, content: JSON.stringify(result) }
    case "gemini":
      return { functionResponse: { ...identified(asked), name: asked.name, response: { output: result } } }
    case "vercel":
      return { type: "tool-result", toolCallId: asked.id, toolName: asked.name, output: { type: "json", value: result } }
    case "mcp":
      return {
        content: [{ type: "text", text: JSON.stringify(result) }],
        ...(isObject(result) ? { structuredContent: result } : {}),
      }
  }
}

function failure(format: ToolFormat, asked: Asked, error: ToolError): ToolResults[ToolFormat] {
  const payload = { error: { code: error.code, message: error.message } }
  switch (format) {
    case "anthropic":
      return { type: "tool_result", tool_use_id: asked.id, content: JSON.stringify(payload), is_error: true }
    case "openai":
      return { type: "function_call_output", call_id: asked.id, output: JSON.stringify(payload) }
    case "openai-chat":
      return { role: "tool", tool_call_id: asked.id, content: JSON.stringify(payload) }
    case "gemini":
      return { functionResponse: { ...identified(asked), name: asked.name, response: payload } }
    case "vercel":
      return {
        type: "tool-result",
        toolCallId: asked.id,
        toolName: asked.name,
        // The AI SDK has no tool-error content part: a refusal is a result whose output says it is one.
        output: { type: "error-json", value: payload },
      }
    case "mcp":
      return { content: [{ type: "text", text: JSON.stringify(payload) }], isError: true }
  }
}

function identified(asked: Asked): { id?: string } {
  return asked.id === "" ? {} : { id: asked.id }
}

// `tool_create` refuses a schema whose root is not an object, so the literal a provider's own type asks for holds.
function objectSchema(schema: JsonSchema): ObjectSchema {
  return { ...schema, type: "object" }
}

function open(schema: JsonSchema): ProviderSchema {
  return { ...schema }
}

// OpenAI structured outputs holds a strict schema to a closed object whose every property is required.
function strictly(schema: JsonSchema): boolean {
  return schema.type === "object" && qualifies(schema, 1)
}

function qualifies(schema: JsonSchema, depth: number): boolean {
  if (depth > STRICT_MAX_DEPTH) return false
  // The supported types are string, number, boolean, integer, object, array, enum and anyOf. A null is none of them.
  if (schema.type === "null") return false
  if (Object.keys(schema).some((keyword) => STRICT_UNSUPPORTED.has(keyword))) return false
  if (schema.type === "object") {
    if (schema.additionalProperties !== false) return false
    const required = new Set(schema.required ?? [])
    if (Object.keys(schema.properties ?? {}).some((key) => !required.has(key))) return false
  }
  if (Object.values(schema.properties ?? {}).some((child) => !qualifies(child, depth + 1))) return false
  return schema.items === undefined || qualifies(schema.items, depth + 1)
}

function subset(schema: JsonSchema): GeminiSchema {
  const kept: Record<string, unknown> = {}
  for (const [keyword, value] of Object.entries(schema)) {
    if (GEMINI_UNSUPPORTED.has(keyword)) continue
    if (keyword === "properties") {
      kept[keyword] = Object.fromEntries(
        Object.entries(value as Record<string, JsonSchema>).map(([child, one]) => [child, subset(one)]),
      )
      continue
    }
    if (GEMINI_COUNTS.has(keyword)) {
      kept[keyword] = String(value)
      continue
    }
    kept[keyword] = keyword === "items" ? subset(value as JsonSchema) : value
  }
  // A genai Schema.enum is string[], so an enum of anything else moves into the description, where a model still reads it.
  if (schema.enum !== undefined && !(schema.type === "string" && schema.enum.every((one) => typeof one === "string"))) {
    delete kept["enum"]
    const values = schema.enum.map((one) => JSON.stringify(one)).join(", ")
    kept["description"] = `${schema.description === undefined ? "" : `${schema.description} `}One of: ${values}.`
  }
  return kept as GeminiSchema
}

function parsed(args: unknown): JsonValue {
  if (typeof args !== "string") return (args ?? {}) as JsonValue
  // Chat Completions sends "" for a call with no arguments at all.
  if (args.trim() === "") return {}
  try {
    return JSON.parse(args) as JsonValue
  } catch (cause) {
    throw new ToolError(
      "invalid_arguments",
      `The arguments are not JSON: ${causeMessage(cause)}. Send the arguments again as one JSON object.`,
    )
  }
}

function object(value: unknown, what: string): Record<string, unknown> {
  if (!isObject(value)) {
    throw new ToolError("invalid_arguments", `${what} is not an object, so there is no tool call to run.`)
  }
  return value
}

function string(value: unknown, field: string): string {
  if (typeof value !== "string" || value === "") {
    throw new ToolError("invalid_arguments", `The tool call carries no \`${field}\`, so there is nothing to run.`)
  }
  return value
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value)
}

function refusal(cause: unknown): ToolError {
  return cause instanceof ToolError ? cause : new ToolError("internal_error", causeMessage(cause))
}

File: python/src/flintd/formats.py
from __future__ import annotations

import json
import re
from typing import Any, Literal

from .errors import ToolError

Format = Literal["anthropic", "openai", "openai-chat", "gemini"]

FORMATS: tuple[Format, ...] = ("anthropic", "openai", "openai-chat", "gemini")

META_TOOL_NAMES = frozenset(
    {"tool_create", "tool_update", "tool_read", "tool_find", "tool_run", "tool_retire", "tool_history"}
)

TOOL_PREFIX = "fl_"

# A stored name is at most 60 characters of `[a-z][a-z0-9_]*`, so `fl_` still fits the 64 every provider allows.
NAME_RULES: dict[str, re.Pattern[str]] = {
    "anthropic": re.compile(r"^[a-zA-Z0-9_-]{1,128}$"),
    "openai": re.compile(r"^[a-zA-Z0-9_-]{1,64}$"),
    "openai-chat": re.compile(r"^[a-zA-Z0-9_-]{1,64}$"),
    "gemini": re.compile(r"^[a-zA-Z_][a-zA-Z0-9_.:-]{0,63}$"),
}

# Gemini's Schema is a subset of OpenAPI 3.03: it has no `additionalProperties`, and its examples field is singular.
GEMINI_UNSUPPORTED = frozenset({"additionalProperties", "examples"})

# A genai Schema spells these four as strings, the way proto JSON spells an int64. `maximum` and `minimum` stay numbers.
GEMINI_COUNTS = frozenset({"minItems", "maxItems", "minLength", "maxLength"})

# Structured outputs names the keywords it takes, and these five of flintd's are not among them.
STRICT_UNSUPPORTED = frozenset({"minLength", "maxLength", "title", "default", "examples"})

# "Objects have limitations on nesting depth and size": up to 10 levels. flintd's own bound is 32.
STRICT_MAX_DEPTH = 10


def export_name(name: str) -> str:
    return name if name in META_TOOL_NAMES else f"{TOOL_PREFIX}{name}"


def library_name(exported: str) -> str:
    """`fl_` is reserved at create, so a Tool never carries it and the prefix is only the spelling a format exports."""
    if exported in META_TOOL_NAMES or not exported.startswith(TOOL_PREFIX):
        return exported
    bare = exported[len(TOOL_PREFIX) :]
    return exported if bare in META_TOOL_NAMES else bare


def format_tools(tools: list[dict[str, Any]], format: Format) -> list[dict[str, Any]]:
    """The `tools` of one provider's request, built from the model-facing list the daemon answers."""
    if format == "anthropic":
        return [
            {"name": named(one, format), "description": one["description"], "input_schema": one["parameters"]}
            for one in tools
        ]
    if format == "openai":
        return [
            {
                "type": "function",
                "name": named(one, format),
                "description": one["description"],
                "parameters": one["parameters"],
                "strict": strictly(one["parameters"]),
            }
            for one in tools
        ]
    if format == "openai-chat":
        return [
            {
                "type": "function",
                "function": {
                    "name": named(one, format),
                    "description": one["description"],
                    "parameters": one["parameters"],
                    "strict": strictly(one["parameters"]),
                },
            }
            for one in tools
        ]
    if format == "gemini":
        return [
            {
                "name": named(one, format),
                "description": one["description"],
                "parameters": subset(one["parameters"]),
                **({} if one.get("result") is None else {"response": subset(one["result"])}),
            }
            for one in tools
        ]
    raise unknown(format)


def read_call(block: Any, format: Format) -> tuple[str, str, Any]:
    """The id the provider wants echoed, the Tool name and the arguments, read from one tool-call block."""
    one = obj(block, "the tool call")
    if format == "anthropic":
        return string(one.get("id"), "id"), string(one.get("name"), "name"), one.get("input")
    if format == "openai":
        return string(one.get("call_id"), "call_id"), string(one.get("name"), "name"), one.get("arguments")
    if format == "openai-chat":
        call = obj(one.get("function"), "function")
        return string(one.get("id"), "id"), string(call.get("name"), "name"), call.get("arguments")
    if format == "gemini":
        part = obj(one["functionCall"], "functionCall") if "functionCall" in one else one
        held = part.get("id")
        return "" if held is None else string(held, "id"), string(part.get("name"), "name"), part.get("args")
    raise unknown(format)


def success(call_id: str, name: str, result: Any, format: Format) -> dict[str, Any]:
    """The block a caller puts back into the provider's next request, holding the result as JSON."""
    if format == "anthropic":
        return {"type": "tool_result", "tool_use_id": call_id, "content": json.dumps(result)}
    if format == "openai":
        return {"type": "function_call_output", "call_id": call_id, "output": json.dumps(result)}
    if format == "openai-chat":
        return {"role": "tool", "tool_call_id": call_id, "content": json.dumps(result)}
    if format == "gemini":
        return {"functionResponse": {**identified(call_id), "name": name, "response": {"output": result}}}
    raise unknown(format)


def failure(call_id: str, name: str, cause: ToolError, format: Format) -> dict[str, Any]:
    """The same block carrying a refusal, in the slot each provider keeps for one, so the model reads it and acts."""
    payload = {"error": {"code": cause.code, "message": cause.message}}
    if format == "anthropic":
        return {"type": "tool_result", "tool_use_id": call_id, "content": json.dumps(payload), "is_error": True}
    if format == "openai":
        return {"type": "function_call_output", "call_id": call_id, "output": json.dumps(payload)}
    if format == "openai-chat":
        return {"role": "tool", "tool_call_id": call_id, "content": json.dumps(payload)}
    if format == "gemini":
        return {"functionResponse": {**identified(call_id), "name": name, "response": payload}}
    raise unknown(format)


def parsed(args: Any) -> Any:
    if not isinstance(args, str):
        return {} if args is None else args
    try:
        return json.loads(args)
    except json.JSONDecodeError as cause:
        raise ToolError(
            "invalid_arguments", f"The arguments are not JSON: {cause}. Send the arguments again as one JSON object."
        ) from None


def identified(call_id: str) -> dict[str, str]:
    return {} if call_id == "" else {"id": call_id}


def named(tool: dict[str, Any], format: Format) -> str:
    name = export_name(tool["name"])
    if NAME_RULES[format].match(name) is None:
        raise ToolError(
            "internal_error", f"{name} is not a name the {format} format accepts.", {"name": name, "format": format}
        )
    return name


def strictly(schema: Any) -> bool:
    """OpenAI structured outputs holds a strict schema to a closed object whose every property is required."""
    return isinstance(schema, dict) and schema.get("type") == "object" and qualifies(schema, 1)


def qualifies(schema: Any, depth: int) -> bool:
    if not isinstance(schema, dict) or depth > STRICT_MAX_DEPTH:
        return False
    # The supported types are string, number, boolean, integer, object, array, enum and anyOf. A null is none of them.
    if schema.get("type") == "null":
        return False
    if any(key in STRICT_UNSUPPORTED for key in schema):
        return False
    properties: dict[str, Any] = schema.get("properties") or {}
    if schema.get("type") == "object":
        if schema.get("additionalProperties") is not False:
            return False
        if any(key not in (schema.get("required") or []) for key in properties):
            return False
    if any(not qualifies(child, depth + 1) for child in properties.values()):
        return False
    return schema.get("items") is None or qualifies(schema["items"], depth + 1)


def subset(schema: Any) -> Any:
    if not isinstance(schema, dict):
        return schema
    kept: dict[str, Any] = {}
    for key, value in schema.items():
        if key in GEMINI_UNSUPPORTED:
            continue
        if key == "properties" and isinstance(value, dict):
            kept[key] = {child: subset(one) for child, one in value.items()}
            continue
        if key in GEMINI_COUNTS:
            kept[key] = str(value)
            continue
        kept[key] = subset(value) if key == "items" else value
    held = schema.get("enum")
    # A genai Schema.enum is string[], so an enum of anything else moves into the description, where a model reads it.
    if held is not None and not (schema.get("type") == "string" and all(isinstance(one, str) for one in held)):
        kept.pop("enum", None)
        values = ", ".join(json.dumps(one) for one in held)
        said = schema.get("description")
        kept["description"] = f"{'' if said is None else f'{said} '}One of: {values}."
    return kept


def unknown(format: str) -> ToolError:
    return ToolError("invalid_arguments", f"The tool format {format!r} is not one of {', '.join(FORMATS)}.")


def obj(value: Any, what: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ToolError("invalid_arguments", f"{what} is not an object, so there is no tool call to run.")
    return value


def string(value: Any, field: str) -> str:
    if not isinstance(value, str) or value == "":
        raise ToolError("invalid_arguments", f"The tool call carries no `{field}`, so there is nothing to run.")
    return value
